File: DSA/SortingAlgorithms/Code/BubbleSort.py
# Bubble Sort in Python
# Repeatedly swap adjacent elements if they are in wrong order

def bubble_sort(arr):
    """Basic bubble sort"""
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

def bubble_sort_optimized(arr):
    """Optimized bubble sort with early termination"""
    n = len(arr)
    for i in range(n - 1):
        swapped = False
        for j in range(n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True
        if not swapped:
            break
    return arr

def bubble_sort_descending(arr):
    """Bubble sort in descending order"""
    n = len(arr)
    for i in range(n - 1):
        for j in range(n - i - 1):
            if arr[j] < arr[j + 1]:  # Changed comparison for descending
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
    return arr

# Test 1: Basic bubble sort
arr = [64, 34, 25, 12, 22, 11, 90]
print("Original Array:", arr)
result = bubble_sort(arr.copy())
print("Sorted (Bubble Sort):", result)

# Test 2: Optimized version
arr2 = [64, 34, 25, 12, 22, 11, 90]
result2 = bubble_sort_optimized(arr2)
print("Sorted (Optimized):", result2)

# Test 3: Already sorted
arr3 = [1, 2, 3, 4, 5]
print("\nAlready sorted:", arr3)
result3 = bubble_sort_optimized(arr3.copy())
print("After Bubble Sort:", result3)

# Test 4: Reverse sorted
arr4 = [5, 4, 3, 2, 1]
print("\nReverse sorted:", arr4)
result4 = bubble_sort(arr4.copy())
print("After Bubble Sort:", result4)

# Test 5: Descending order
arr5 = [64, 34, 25, 12, 22, 11, 90]
print("\nDescending order sorting:")
result5 = bubble_sort_descending(arr5.copy())
print("Sorted (Descending):", result5)

# Test 6: Array with duplicates
arr6 = [5, 2, 8, 2, 9, 1, 5, 5]
print("\nArray with duplicates:", arr6)
result6 = bubble_sort_optimized(arr6.copy())
print("After Bubble Sort:", result6)

print("\nComplexity Analysis:")
print("Time Complexity: O(n^2) - Quadratic")
print("Space Complexity: O(1) - In-place")
print("Best Case: O(n) - Already sorted (optimized)")
print("Worst Case: O(n^2) - Reverse sorted")
print("Stable: Yes")

File: DSA/SortingAlgorithms/Code/BubbleSort.cpp
#include <iostream>
#include <vector>
using namespace std;

// Bubble Sort in C++
// Repeatedly swap adjacent elements if they are in wrong order

void bubbleSort(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
            }
        }
    }
}

// Optimized Bubble Sort with early termination
void bubbleSortOptimized(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n - 1; i++) {
        bool swapped = false;
        for (int j = 0; j < n - i - 1; j++) {
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                swapped = true;
            }
        }
        if (!swapped) break;
    }
}

void displayArray(const vector<int>& arr) {
    for (int num : arr)
        cout << num << " ";
    cout << "\n";
}

int main() {
    vector<int> arr = {64, 34, 25, 12, 22, 11, 90};
    
    cout << "Original Array: ";
    displayArray(arr);
    
    bubbleSortOptimized(arr);
    
    cout << "Sorted Array (Optimized Bubble Sort): ";
    displayArray(arr);
    
    // Test with already sorted array
    vector<int> arr2 = {1, 2, 3, 4, 5};
    cout << "\nAlready sorted array: ";
    displayArray(arr2);
    bubbleSortOptimized(arr2);
    cout << "After Bubble Sort: ";
    displayArray(arr2);
    
    // Test with reverse sorted array
    vector<int> arr3 = {5, 4, 3, 2, 1};
    cout << "\nReverse sorted array: ";
    displayArray(arr3);
    bubbleSortOptimized(arr3);
    cout << "After Bubble Sort: ";
    displayArray(arr3);
    
    // Test with duplicates
    vector<int> arr4 = {5, 2, 8, 2, 9, 1, 5, 5};
    cout << "\nArray with duplicates: ";
    displayArray(arr4);
    bubbleSortOptimized(arr4);
    cout << "After Bubble Sort: ";
    displayArray(arr4);
    
    cout << "\nComplexity Analysis:\n";
    cout << "Time Complexity: O(n^2) - Quadratic\n";
    cout << "Space Complexity: O(1) - In-place\n";
    cout << "Best Case: O(n) - Already sorted (optimized)\n";
    cout << "Worst Case: O(n^2) - Reverse sorted\n";
    cout << "Stable: Yes\n";
    
    return 0;
}
